sdks/python/apache_beam/runners/worker/dataflow_element_execution_tracker.pyx

```diff
@@ -0,0 +1,330 @@
+    cdef unordered_map[ElementExecutionPtr, int64_t].iterator it = (
+      executions_per_element.begin())
+    cdef ElementExecutionPtr element_ptr = NULL
+    cdef int64_t attribution_time = 0
+    # Attribute processing time.
+    while it != executions_per_element.end():
+      element_ptr = dref(it).first
+      attribution_time = (sample_time / total_execution
+                         * executions_per_element[element_ptr])
+      self.execution_duration[element_ptr] += attribution_time
+      iref(it)
+
+  cdef void update_counter_cache(self, int64_t latest_snapshot,
+      unordered_map[CharPtr, vector[int64_t]]& counter_cache) nogil:
+    """Update counter_cache with values of done elements up to the given 
+    snapshot and prune done_journals.
+    """
+    cdef SnapshottedExecutionPtr snapshotted_ptr = NULL
+    cdef ElementExecutionPtr element_ptr = NULL
+    while self.shared_state.done_journal.journal.empty() is False:
+      snapshotted_ptr = self.shared_state.done_journal.journal.front()
+      if snapshotted_ptr.snapshot <= latest_snapshot:
+        self.shared_state.done_journal.journal.pop_front()
+        element_ptr = snapshotted_ptr.execution_ptr
+        # Unit of sampling time is nano secs, we update counter_cache with ms.
+        (counter_cache[element_ptr.operation_name]
+         .push_back(self.execution_duration[element_ptr]/1000000))
+      else:
+        break
+
+
+cdef class ExecutionJournalWriter(object):
+  """Writes journal entries on element processing state changes.
+
+  Attributes:
+    shared_state: Shared journals with ExecutionJournalReader.
+    execution_stack: Execution stack of processing elements. Elements are pushed
+    onto the stack when scoped_process_state.__enter__ and popped on
+    scoped_process_state.__exit__. This stack mirrors the actual runtime stack
+    and contains the step + element context in order to attribute sampled
+    execution time.
+
+  Args:
+    shared_state: Initialized in DataflowElementExecutionTracker.
+  """
+  def __init__(self, ReaderWriterState shared_state):
+    self.shared_state = shared_state
+    self.execution_stack = deque[ElementExecutionPtr]()
+    self.add_execution(NULL)
+
+  cdef void add_execution(self, ElementExecutionPtr execution_ptr):
+    """Push current ElementExecution onto top of the execution_stack and add
+    new entry into journal.
+    """
+    cdef int64_t next_snapshot = self.shared_state.latest_snapshot + 1
+    self.execution_stack.push_back(execution_ptr)
+    self.shared_state.execution_journal.add_journal(execution_ptr,
+                                                    next_snapshot)
+    self.shared_state.latest_snapshot = next_snapshot
+
+  cdef void start_processing(self, char* operation_name):
+    """Create and journal a new ElementExecution to track a processing element.
+    """
+    cdef ElementExecutionPtr execution_ptr = (
+      <ElementExecutionPtr> malloc(sizeof(ElementExecution)))
```

Comment:
<!--thread_id:cc_182317472_t; commit:bdfbf7813f29cfda39312640547d26ca30bbc5eb; resolved:0-->
<!--section:context-quote-->
> **aaltay** wrote:
> I think you can still use `ElementExecutionPtr` by typedefing it to `char*`, would it work?
> 
> Let's not do 3 in this PR. One can do malloc's to extend the size of this large object when needed. This will happen much less frequently than individually allocation memory for each element.

<!--section:body-->
I'm afraid  ```typedef ElementExecutionPtr char*``` will not work, unless every time we ```malloc``` a new ```char*``` and do ```memcpy``` op_name to this new address.

Reply:
I am a bit confused here.

Currently: ElementExecution is a struct that has a single char*. We do not take ownership of the char*, or neither we need to memcpy contents of this char*. We malloc a new ElementExecution.

If instead we use typedef ElementExecutionPtr, we can exactly do the same thing. We still do not need to take ownership of char* or memcpy it.

Could you explain why do we need to do the copying in the latter case?

Reply:
Are you asking, why we need to create a new object to wrap char* here? If so, we use a new object to mark one element(based on different address, we use this address to group the same element later on), since it's possible that there are many elements going through one operation. We want to calculate processing time per element per operation.

Reply:
Maybe we can have video chat to settle this out.

Yes, my question is why do we need to create a new object. It seems like the char* itself is enough to uniquely identify the element.